### DISCUSSION TITLE
Discrete collections of Entities? "Instances" of the whole ECS?

### DISCUSSION THREAD

**Discussion by transmutrix** [rlipsc/polymorph#28] [Q&A] [answered]
Hey, this library seems pretty rad! I started writing my own ECS in Nim, searched to see what others had done, and found this. It looks so good I may just use it instead of writing my own! It does a lot right philosophically IMO, and lines up with how I'd prefer to write components and systems—without inheritance, sluggish queries, and other common foibles and pain points.

I've been reading through the documentation and I had a question:

Is it possible to have more than one "world" of storage for the ECS, cleanly separated, and pass it into an update or something? That is, if I have two scenes I want loaded in memory at the same time, can I have the entities and components for each live in their own data structure and update them separately?

Currently in my project, using my own non-ECS scheme that I would ultimately replace with polymorph, I have Scenes and I can load as many as I like, and update them at will. I rely on this ability primarily for slide transitions between rooms for now, but I could imagine other uses in the future (cutscenes showing a far away place, minigames that are self-contained, etc).

Thank you for your time!

**Comment by transmutrix** [comment]
After doing more reading, and going through the spaceshooter example, it seems like maybe I could do something like this?

- Make components KeepBetweenScenes and UnloadingScene.
- Make a system `markUnload` that, when unpaused, will flag every Entity `[not KeepBetweenScenes]` by adding UnloadingScene to it, then pause itself.
- Make another system `finishUnload` that, when unpaused, will `finish: sys.clear` for all `[UnloadingScene]`, then pause itself.

Then:

1. When a scene transition is triggered, unpause `sysMarkUnload` to flag existing entities.
2. Load the new scene's entities into the mix.
3. Run the visual transition including both live scenes' entities.
4. When the transition finishes, unpause `sysFinishUnload` to finish actually removing stuff from the old scene.

Does this seem like a reasonable way to handle this use case in polymorph? Are there any obvious caveats I should watch out for?

**Comment by transmutrix** [comment, marked as answer]
So for now I ended up doing the following:

- Keep a list of the entities in a scene outside the ECS system, in my existing scene structure.
- When a scene will unload, add an Unloading component to all the appropriate entities from that list.
- When loading a new scene, mark all the new entities as Loading.
- When the scene transition finishes, systems are unpaused to do necessary cleanup.

All in all, getting a basic thing going, and replacing several Actors' rendering code with ECS Entities with a few components in polymorph instead, took a couple hours of messing about, and well under 100 lines of code. Nice!

I will continue experimenting, and see how far I can take this hybrid approach along the road to a simpler life. :)

Once again, thank you for all your hard work.

**Comment by rlipsc** [comment]
Hey @transmutrix, thanks for the kind words, and I'm glad you've been able to solve this within a singular ECS. 

Polymorph does allow you to create separate ECS "identities", but that's more for defining multiple ECS frameworks at the same time. They can't directly interact, for example, because each has its own distinct `EntityRef` type, and so on.

It sounds like you've got a good solution!

Using components as clear signals with `Loading` and `Unloading` makes a lot of sense. It lets you init/deinit stuff in a defined system order, and I can imagine systems using `[not Loading]` and `[not Unloading]` for various situations.

You can also use events to manage resources if you want instant updates. A toy example:

```nim
import polymorph, tables

register defaultCompOpts: string

makeSystem "NameDb", [string]:
  fields: db: Table[string, EntityRef]
  added: sys.db[string] = entity
  removed: sys.db.del string

makeEcs()

let
  # Creation and construction immediately update "NameDb"
  e1 = newEntity "Hello"
  blueprint = cl("World")
  e2 = blueprint.construct

assert sysNameDb.db["Hello"] == e1
assert sysNameDb.db["World"] == e2
```

There's loads of different types of events that should cover most situations, including things like `onClone` and `onConstruct`.
This could let you manage resources, data structures, and so on at the point of component add/remove.

That said, you may find it better to keep doing things in the system workflow as you are now, as there's no ambiguity to when a load/unload relevant system is running and batching stuff is generally better for performance.

Another option is to use component lists to store other scenes as pure data and spawn them into live entities as needed.
This is fairly ergonomic, as you can do `entities.toTemplate` to create a construction template, pop that in some variable somewhere, then `construct` it when the scene becomes current.

In case you find it useful, an example of loading/saving component lists with streams is demonstrated here: https://github.com/rlipsc/polymorph/discussions/25

**Comment by transmutrix** [comment]
Ah, thank you for the detailed response, and the further reading!
